server/src/bge/System.cpp

```diff
@@ -423,26 +441,150 @@ namespace bge {
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS) {
+                            // effect time = (r-x)^2 * mt / (r^2)
+                            // where r is the radius of the explosion (max distance away where players are still affected)
+                            // x is the distance of this player from the explosion
+                            // mt is the maximum amount of time the effect can last
+                            // (So the effect will last the maximum amount of time for players who are exactly at the ball and 0 time for players who are exactly at the boundary)
+                            statusEffects.movementSpeedTicksLeft = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_PROJ_EFFECT_LENGTH / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                            statusEffects.alternateMovementSpeed = SLOW_MOVEMENT_SPEED;
+                        }
+                    }
+                    else if (projData.type == SPRING) {
+                        // We may want to use different constants for explosion radius in the future for different seasons,
+                        // so the radius check is separated by season
+                        // Healing yourself by right clicking the ground as often as possible seems to OP,
+                        // so the health effect only works on other players
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS && projData.creatorId != playerEntity.id) {
+                            float healStrength = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_HEAL_STRENGTH / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                            health.healthPoint += healStrength;
+                            if (health.healthPoint > PLAYER_MAX_HEALTH) {
+                                health.healthPoint = PLAYER_MAX_HEALTH;
+                            }
+                        }
+                    }
+                    else if (projData.type == SUMMER) {
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS) {
+                            statusEffects.swappedControlsTicksLeft = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_PROJ_EFFECT_LENGTH / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                        }
+                    }
+                    else if (projData.type == AUTUMN) {
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS) {
+                            glm::vec3 launchDir = glm::normalize(playerPos.position - pos.position);
+                            float launchSeverity = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_LAUNCH_SEVERITY / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                            playerVel.velocity += launchSeverity * launchDir;
```

Comment:
i tried out winter, spring and summer's abilities, they're so cool:) !

when I throw the autumn ball, I somehow get teleported immediately into the void.... my velocity is printed (nan,nan,nan)

what might happen:
- the autumn projectile collides immediately with its thrower
- thus `playerPos.position - pos.position` is a zero vector
- normalizing a zero vector will gg

fix: just launch the projectile from outside the player (I think you guys already did this but didnt update projPos.position )

Reply:
What exactly did you do? I can't reproduce the bug with autumn you're describing (I haven't pulled your latest commit).